src/Microsoft.Health.SqlServer/Features/Client/SqlConnectionWrapper.cs

```diff
@@ -17,55 +17,72 @@ public class SqlConnectionWrapper : IDisposable
         private readonly SqlCommandWrapperFactory _sqlCommandWrapperFactory;
+        private readonly ISqlConnectionFactory _sqlConnectionFactory;
 
-        public SqlConnectionWrapper(
-            SqlServerDataStoreConfiguration configuration,
+        private SqlConnection _sqlConnection;
+        private SqlTransaction _sqlTransaction;
+
+        internal SqlConnectionWrapper(
             SqlTransactionHandler sqlTransactionHandler,
             SqlCommandWrapperFactory sqlCommandWrapperFactory,
+            ISqlConnectionFactory sqlConnectionFactory,
             bool enlistInTransactionIfPresent)
         {
-            EnsureArg.IsNotNull(configuration, nameof(configuration));
             EnsureArg.IsNotNull(sqlTransactionHandler, nameof(sqlTransactionHandler));
             EnsureArg.IsNotNull(sqlCommandWrapperFactory, nameof(sqlCommandWrapperFactory));
+            EnsureArg.IsNotNull(sqlConnectionFactory, nameof(sqlConnectionFactory));
 
             _sqlTransactionHandler = sqlTransactionHandler;
             _enlistInTransactionIfPresent = enlistInTransactionIfPresent;
             _sqlCommandWrapperFactory = sqlCommandWrapperFactory;
+            _sqlConnectionFactory = sqlConnectionFactory;
+        }
 
-            if (_enlistInTransactionIfPresent && sqlTransactionHandler.SqlTransactionScope?.SqlConnection != null)
+        public SqlConnection SqlConnection
+        {
+            get { return _sqlConnection; }
+        }
+
+        public SqlTransaction SqlTransaction
+        {
+            get
             {
-                SqlConnection = sqlTransactionHandler.SqlTransactionScope.SqlConnection;
+                return _sqlTransaction;
+            }
+        }
+
+        internal async Task InitializeAsync()
+        {
+            if (_enlistInTransactionIfPresent && _sqlTransactionHandler.SqlTransactionScope?.SqlConnection != null)
+            {
+                _sqlConnection = _sqlTransactionHandler.SqlTransactionScope.SqlConnection;
             }
             else
             {
-                SqlConnection = new SqlConnection(configuration.ConnectionString);
+                _sqlConnection = await _sqlConnectionFactory.GetSqlConnectionAsync();
             }
 
-            if (_enlistInTransactionIfPresent && sqlTransactionHandler.SqlTransactionScope != null && sqlTransactionHandler.SqlTransactionScope.SqlConnection == null)
+            if (_enlistInTransactionIfPresent && _sqlTransactionHandler.SqlTransactionScope != null && _sqlTransactionHandler.SqlTransactionScope.SqlConnection == null)
             {
-                sqlTransactionHandler.SqlTransactionScope.SqlConnection = SqlConnection;
+                _sqlTransactionHandler.SqlTransactionScope.SqlConnection = SqlConnection;
             }
 
             if (SqlConnection.State != ConnectionState.Open)
             {
-                SqlConnection.Open();
+                await SqlConnection.OpenAsync();
```

Comment:
Most of the Async calls takes CancellationToken so we need to honor them.

Reply:
A fix for this is made in [AB#106](https://github.com/microsoft/healthcare-shared-components/pull/106).